### DISCUSSION TITLE
RFID music player for kids

### DISCUSSION THREAD

**Discussion by 0lek** [micro-nova/AmpliPi#389] [Show and tell] [answered]
Hey all,

I'm currently working on an RFID music player for my daughter (she just turned 4). For those who don't understand the concept: Instead of using a screen to select the song you want (which is kinda hard if you can't read yet + the less she looks at a screen, the better), you just use an RFID card. 

So the user story for this is: pick a card, hold it to the reader and that song begins to play. Hold another one to it and another song plays. No queues and other bullshit ;)

I first tried to do it with a bluetooth speaker. It worked but I couldn't make it reliable (as in it works after a power cycle - the bluetooth speaker had to be paired using the CLI). The other option was to use some kind of audio hat (e.g. HiFiBerry). However, as I have an amplipi, that's probably the easiest route. It requires only a Raspberry (I use a zero) and an RFID card reader (RC522). The script is pretty simple, it uses pyamplipi and the announce API.

With the newest version (with LMS client) I will see whether I can port it to use LMS instead of the announce API (seems better suited). I do not have an LMS server, but maybe I can run it simply on the pi for now? Both LMS and announce have the advantage of easily using local files (or network files) instead of streams. Spotify seems to be too much overhead coding wise (as in select playlist, select song - not sure if this is even possible in the API). I don't have much time for tinkering :)

@linknum23 Do you think using either announce or LMS is the way to go?

To dos:

- Port to use LMS
- Make it automatically start the script after booting
- Figure out power (battery would be ideal)
- Figure out hardware buttons to select zones (as in her room or the living room)
- Figure out volume control

This might take a while. The MVP works. The funny thing is, I can even use a credit card for launching songs :)

Code below:
(BTW, the dropbox one doesn't work)

```
#! /usr/bin/env python3

import asyncio
import sys, os, command
import RPi.GPIO as GPIO
from collections.abc import Iterable
from pyamplipi.amplipi import AmpliPi
from pyamplipi.models import Announcement, Status
from mfrc522 import SimpleMFRC522

# Set up some important stuff
reader = SimpleMFRC522()
loop = asyncio.new_event_loop()

async def main():
    loop = asyncio.get_event_loop()
    GPIO.setwarnings(False)
    print('Show me your card!')
    while True:
        # Get ID of the RFID card
        id, text = reader.read()
        print('I have read this card:', id)
        print(type(id))
        if id == 323511587440:
            print('I have actually recognized the card')
            uri = "https://file-examples.com/storage/feb2e515cc6339d7ba1ffcd/2017/11/file_example_MP3_700KB.mp3"
            await announce(uri)
        if id == 849498534822:
            print('Other card')
            uri = "https://www.dropbox.com/s/y5phdjd42mgy6uf/kolko.mp3?dl=0"
            await announce(uri)
        else:
            print('But I\'m stupid')

async def announce(sound_uri):
    api_uri = "http://10.10.10.42/api"
    print(f"Using announce\n  at {api_uri}\n  to play sound at {sound_uri}\n\n")

    sndsrv = AmpliPi(api_uri, timeout=30)
    await play_sound(sndsrv, sound_uri)
    await sndsrv.close()
    print("done")

async def play_sound(sndsrv, sound_uri):
    sound: Announcement  = Announcement(media=sound_uri, vol_f=0.6, source_id=1)
    await sndsrv.announce(sound)

loop.run_until_complete(main())
```

**Comment by linknum23** [comment]
Awesome! I was planning on doing something like this for my 2 year old starting from this post: https://www.home-assistant.io/blog/2020/09/15/home-assistant-tags/

That said announce might be too simple for this use case. It was intended for short clips and doesn't really support an announcement interrupting a previous announcement, something that will happen when you daughter decides to play a different song halfway through. 

For this I would use the stream type that announce uses under the hood: fileplayer. 

If I did this I would create a single file player stream dedicated to this task. Every time a song was requested I would set the streams url to the song url, then play that stream on an open source with the zone requested. For simplicity of determining and open source I would dedicate source 4 to this application. 

As for LMS I do not have any idea of how that API works.

I hope that helps.

**Reply by linknum23** [reply]
Edited the above post. With those tweeks I think this could be pretty simple and awesome.

**Reply by 0lek** [reply]
@linknum23 Thanks! Announce was simply a proof of concept :) I never thought of using tags within HA. It also seems overkill. And I would anyhow need some - I don't know - HA satellite? On a raspi? 

That said, I agree with fileplayer. However - how do I set this up? I can't select fileplayer within the UI (gear --> streams)?

Thx

**Reply by linknum23** [reply]
You need to use the API. ~~Look at the create_stream documentation.~~ Oops it looks like the docs were missing the type! Here's
what I would put in the create_stream request.

{
  "name": "Song Player",
  "type": "fileplayer",
   "url": ""
}

**Reply by 0lek** [reply]
Hey @linknum23 

Still working on this, but wanna wrap up soon. Quick Q: When I try to "play" the stream (via a command) the API gives me: {"detail":"Failed to execute stream command: play: Unicorn Stream does not support commands"}

How do I then play the fileplayer stream?

Thx

**Reply by 0lek** [reply]
Wenn I look at the status response from the API, the fileplayer stream I created also does not show any available commands. So I guess this doesn't work. I can't play this stream even through the UI, so I guess that's that... :(

**Reply by linknum23** [reply, marked as answer]
It should start playing immediately after set_source is called. The file player stream does not support any play or pause commands at the moment since it is almost entirely used for announcements.

I hope this helps. We really should investigate adding a full fledged local file player.

**Reply by 0lek** [reply]
Hey @linknum23 

Thanks. It really does. Works.

And you should just add a local file player and not just investigate ;)

I finally managed to get it working. I made a quick detour into the Logitech Media Server, but this is probably easiest. 

What I don't like about this solution that there's no way to stop the playing song, even via the UI.

Anyways, in case you want to do the same for your daughter, here's the code:

```
#!/usr/bin/env python3

import sys, os
import json

from mfrc522 import SimpleMFRC522

def set_stream(sound_uri):
	# We need to it this way, because using pyamplipi gets a "Method not allowed" from the API
	cmd = "curl -s -o /dev/null -X PATCH 10.10.10.42/api/streams/1005 -H 'Content-Type: application/json' -d '{\"url\":\"" + sound_uri + "\"}'"
	os.system(cmd)

def load_preset():
	cmd = "curl -s -o /dev/null -X POST 10.10.10.42/api/presets/10001/load"
	os.system(cmd)

def main():
	# Set up RFID
	reader = SimpleMFRC522()
	
	# Load preset on AmpliPi
	load_preset()
	
	# Open the cards file
	file = open("/home/pi/unicorn/cards.json", "r")
	content = file.read()
	cards = json.loads(content)

	try:
		while True:
			# Get ID of the RFID card
			print("SHOW YOUR TAG")
			id, text = reader.read()
			print("TAG FOUND: " + str(id))

			# Look for our card
			for card in cards:
				if id == card['id']:
					uri = "http://10.10.10.45/unicorn/" + card['file']
					set_stream(uri)
					print(card['name'])
	except KeyboardInterrupt:
		print("KeyboardInterrupt")

	sys.exit(0)

if __name__ == "__main__":
	main()
```
I run it as a daemon.

Cheers
Aleks

**Reply by linknum23** [reply]
So to just get stop to work you can actually use the internet radio stream instead of the file player. Let me check if that will require a syntax change. I want you to be able to stop toddler music if necessary :smile:

**Reply by linknum23** [reply]
The syntax doesn't change and you can add album art if you want. One caveat is that the system doesn't check for when the file stops playing so it will look like the file is always playing even after playback is stopped.
